drivers/flash/flash_simulator.c

```diff
@@ -0,0 +1,344 @@
+/* increment a unit erase cycles (EC) statistic */
+#define ERASE_CYCLES_INC(U) (*(&flash_sim_stats.erase_cycles_unit0 + (U)) += 1)
+/* increment a unit excess erase cycles (EEC) statistic */
+#define EXCESS_CYCLES_INC(U)                                                   \
+	(*(&flash_sim_stats.excess_erase_cycles_unit0 + (U)) += 1)
+
+/* simulator statistcs */
+
+STATS_SECT_START(flash_sim_stats)
+STATS_SECT_ENTRY32(flash_read_calls)       /* calls to flash_read() */
+STATS_SECT_ENTRY32(flash_read_time_us)     /* time spent in flash_read() */
+STATS_SECT_ENTRY32(flash_read_api_errors)  /* flash_read() API errors */
+STATS_SECT_ENTRY32(flash_read_hw_errors)   /* hardware errors during read */
+STATS_SECT_ENTRY32(flash_write_calls)      /* calls to flash_write() */
+STATS_SECT_ENTRY32(flash_write_time_us)    /* time spent in flash_write() */
+STATS_SECT_ENTRY32(flash_write_api_errors) /* flash_write() API errors */
+STATS_SECT_ENTRY32(flash_write_hw_errors)  /* hardware errors during write */
+STATS_SECT_ENTRY32(flash_erase_calls)      /* calls to flash_erase() */
+STATS_SECT_ENTRY32(flash_erase_time_us)    /* time spent in flash_erase() */
+STATS_SECT_ENTRY32(flash_erase_api_errors) /* flash_erase() API errors */
+STATS_SECT_ENTRY32(flash_erase_hw_errors)  /* hardware errors during erase */
+STATS_SECT_ENTRY32(bytes_written)	   /* total bytes written */
+STATS_SECT_ENTRY32(bytes_read)		   /* total bytes read */
+ /* erase cycle count for unit */
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE, STATS_SECT_EC))
+ /* excess erase cycles for unit */
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE, STATS_SECT_EEC))
+STATS_SECT_END;
+
+STATS_SECT_DECL(flash_sim_stats) flash_sim_stats;
+STATS_NAME_START(flash_sim_stats)
+STATS_NAME(flash_sim_stats, flash_read_calls)
+STATS_NAME(flash_sim_stats, flash_read_time_us)
+STATS_NAME(flash_sim_stats, flash_read_api_errors)
+STATS_NAME(flash_sim_stats, flash_read_hw_errors)
+STATS_NAME(flash_sim_stats, flash_write_calls)
+STATS_NAME(flash_sim_stats, flash_write_time_us)
+STATS_NAME(flash_sim_stats, flash_write_api_errors)
+STATS_NAME(flash_sim_stats, flash_write_hw_errors)
+STATS_NAME(flash_sim_stats, flash_erase_calls)
+STATS_NAME(flash_sim_stats, flash_erase_time_us)
+STATS_NAME(flash_sim_stats, flash_erase_api_errors)
+STATS_NAME(flash_sim_stats, flash_erase_hw_errors)
+STATS_NAME(flash_sim_stats, bytes_written)
+STATS_NAME(flash_sim_stats, bytes_read)
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE, STATS_NAME_EC))
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE, STATS_NAME_EEC))
+STATS_NAME_END(flash_sim_stats);
+
+static u8_t mock_flash[FLASH_SIZE];
+static bool write_protection;
+static u64_t rand_seed;
+
+static const struct flash_driver_api flash_sim_api;
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+static int toss(int failure_rate_pm)
+{
+	rand_seed = 1103515245 * rand_seed + 12345;
+	return ((u32_t)rand_seed % 1000) < failure_rate_pm ? -EIO : 0;
```

Comment:
Not sure how much this matters, but the division is pretty expensive on some platforms.

Reply:
Probably no matters as we mainly intended to us it for <i>qemu</i> simulations test.

Reply:
hw error sim is not in this patch anymore